src/server/pkg/storage/gc/garbage_collector.go

```diff
@@ -0,0 +1,197 @@
+package gc
+
+import (
+	"context"
+	"strconv"
+	"time"
+
+	"github.com/jinzhu/gorm"
+	"github.com/pachyderm/pachyderm/src/server/pkg/obj"
+)
+
+const (
+	defaultPolling = 5 * time.Minute
+	defaultTimeout = 30 * time.Minute
+)
+
+const (
+	polling                     = "polling"
+	deletingTemporaryReferences = "deleting temporary references"
+	checkingChunks              = "checking chunks"
+	markingDeletingChunks       = "marking deleting chunks"
+	deletingChunks              = "deleting chunks"
+	removingChunkRows           = "removing chunk rows"
+)
+
+type garbageCollector struct {
+	objClient        obj.Client
+	db               *gorm.DB
+	polling, timeout time.Duration
+}
+
+// Run runs the garbage collector.
+func Run(ctx context.Context, objClient obj.Client, db *gorm.DB, opts ...Option) error {
+	gc := &garbageCollector{
+		objClient: objClient,
+		db:        db,
+		polling:   defaultPolling,
+		timeout:   defaultTimeout,
+	}
+	for _, opt := range opts {
+		opt(gc)
+	}
+	return gc.pollingFunc(ctx)
+}
+
+func (gc *garbageCollector) maybeDeleteTemporaryRefs(ctx context.Context) error {
+	seconds := strconv.Itoa(int(gc.timeout / time.Second))
+	stmtFuncs := []statementFunc{
+		func(txn *gorm.DB) *gorm.DB {
+			return txn.Exec(`
+				DELETE FROM refs
+				WHERE sourcetype = 'temporary'
+				AND created < NOW() - INTERVAL '` + seconds + `' second
+			`)
+		},
+	}
+	return runTransaction(ctx, gc.db, deletingTemporaryReferences, stmtFuncs)
+}
+
+func (gc *garbageCollector) maybeDeleteChunks(ctx context.Context) error {
+	chunksToDelete := []chunkModel{}
+	stmtFuncs := []statementFunc{
+		func(txn *gorm.DB) *gorm.DB {
+			return txn.Raw(`
+				SELECT chunks.chunk
+				FROM chunks
+				LEFT OUTER JOIN refs
+				ON chunks.chunk = refs.chunk
+				WHERE refs.chunk IS NULL
+			`).Scan(&chunksToDelete)
+		},
+	}
+	if err := runTransaction(ctx, gc.db, checkingChunks, stmtFuncs); err != nil {
+		return err
+	}
+	return gc.deleteChunks(ctx, convertChunks(chunksToDelete))
+}
+
+func (gc *garbageCollector) pollingFunc(ctx context.Context) error {
+	return retry(polling, func() error {
+		for {
+			if err := gc.maybeDeleteTemporaryRefs(ctx); err != nil {
+				return err
+			}
+			if err := gc.maybeDeleteChunks(ctx); err != nil {
+				return err
+			}
+			select {
+			case <-time.After(gc.polling):
+			case <-ctx.Done():
+				return nil
```

Comment:
It seems a little bit weird here is that if `ctx` gets cancelled while one of the calls above is running you're going to get a context deadline exceeded error (I assume, didn't look to closely at those methods), but if it gets cancelled while we're sleeping then you don't get an error. Does that matter? And would it make sense to unify that behavior.

Reply:
This is definitely a problem because we would retry when we shouldn't (when the context is canceled, the polling function should immediately return). I am going to fix this and follow the pattern of returning the context error when waiting on done (this is the pattern we should be following in general).